tools/btmidi-server.c

```diff
@@ -0,0 +1,1032 @@
+					BT_GATT_CHRC_PROP_READ |
+					BT_GATT_CHRC_PROP_EXT_PROP,
+					gap_device_name_read_cb,
+					gap_device_name_write_cb,
+					server);
+
+	bt_uuid16_create(&uuid, GATT_CHARAC_EXT_PROPER_UUID);
+	gatt_db_service_add_descriptor(service, &uuid, BT_ATT_PERM_READ,
+					gap_device_name_ext_prop_read_cb, NULL, server);
+
+	/*
+	 * Appearance characteristic. Reads and writes should obtain the value
+	 * from the database.
+	 */
+	bt_uuid16_create(&uuid, GATT_CHARAC_APPEARANCE);
+	tmp = gatt_db_service_add_characteristic(service, &uuid,
+							BT_ATT_PERM_READ,
+							BT_GATT_CHRC_PROP_READ,
+							NULL, NULL, server);
+
+	/*
+	 * Write the appearance value to the database, since we're not using a
+	 * callback.
+	 */
+	put_le16(128, &appearance);
+	gatt_db_attribute_write(tmp, 0, (void *) &appearance,
+							sizeof(appearance),
+							BT_ATT_OP_WRITE_REQ,
+							NULL, confirm_write,
+							NULL);
+#endif
+	gatt_db_service_set_active(service, true);
+}
+
+static void populate_gatt_service(struct server *server)
+{
+	bt_uuid_t uuid;
+	struct gatt_db_attribute *service, *svc_chngd;
+
+	/* Add the GATT service */
+	bt_uuid16_create(&uuid, UUID_GATT);
+	service = gatt_db_add_service(server->db, &uuid, true, 4);
+
+#if ENABLE_EXTRA_CHARACTERISTICS
+	bt_uuid16_create(&uuid, GATT_CHARAC_SERVICE_CHANGED);
+	svc_chngd = gatt_db_service_add_characteristic(service, &uuid,
+			BT_ATT_PERM_READ,
+			BT_GATT_CHRC_PROP_READ | BT_GATT_CHRC_PROP_INDICATE,
+			gatt_service_changed_cb,
+			NULL, server);
+	server->gatt_svc_chngd_handle = gatt_db_attribute_get_handle(svc_chngd);
+
+	bt_uuid16_create(&uuid, GATT_CLIENT_CHARAC_CFG_UUID);
+	gatt_db_service_add_descriptor(service, &uuid,
+				BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+				gatt_svc_chngd_ccc_read_cb,
+				gatt_svc_chngd_ccc_write_cb, server);
+#endif
+	gatt_db_service_set_active(service, true);
+}
+
+static int populate_midi_service(struct server *server, int client, int port)
+{
+	bt_uuid_t uuid;
+	struct gatt_db_attribute *service;
+	struct pollfd pfd;
+	int err;
+	snd_seq_client_info_t *info;
+
+	/* ALSA Sequencer Client and Port Setup */
+	err = snd_seq_open(&server->seq_handle, "default", SND_SEQ_OPEN_DUPLEX, 0);
+	if (err < 0) {
+		fprintf(stderr, "Could not open ALSA Sequencer: %s (%d)\n", snd_strerror(err), err);
+		return 1;
+	}
+
+	err = snd_seq_nonblock(server->seq_handle, SND_SEQ_NONBLOCK);
+	if (err < 0) {
+		fprintf(stderr, "Could not set nonblock mode: %s (%d)\n", snd_strerror(err), err);
+		goto _err_handle;
+	}
+
+	err = snd_seq_set_client_name(server->seq_handle, gattName);
+	if (err < 0) {
+		fprintf(stderr, "Could not configure ALSA client: %s (%d)\n", snd_strerror(err), err);
+		goto _err_handle;
+	}
+
+	err = snd_seq_client_id(server->seq_handle);
+	if (err < 0) {
+		fprintf(stderr, "Could not retrieve ALSA client: %s (%d)\n", snd_strerror(err), err);
+		goto _err_handle;
+	}
+	server->seq_client_id = err;
+
+	err = snd_seq_create_simple_port(server->seq_handle, gattName,
```

Comment:
Hmm, ok.
I think we should structure the program in a more asynchronous way, since this is a type of asynchronous network programming.
For example, setup all the resources first. Then accept for a incoming connection, setup resources for connection and then send/read MIDI data. Cleanup at the end.